File: src/Adj6ToGCSR-lt-reorder2 (under construction)/stage1.go
package main

import (
	"os"
	"path/filepath"
)

func stage1_mapper(adj6 []uint8, in <-chan sRawDat, gWidth uint64) <-chan []gridEdge {
	out := make(chan []gridEdge, 128)

	go func() {
		defer close(out)
		for dat := range in {
			el := make([]gridEdge, dat.cnt)
			selfloop := uint64(0)

			for i := uint64(0); i < dat.cnt; i++ {
				s := dat.src
				now := dat.dstStart + i*6
				d := be6_le8(adj6[now : now+6])

				if s < d {
					s, d = d, s
				} else if s == d {
					selfloop++
					continue
				}

				el[i-selfloop] = gridEdge{
					gidx: [2]uint32{uint32(s / gWidth), uint32(d / gWidth)},
					edge: [2]uint32{uint32(s % gWidth), uint32(d % gWidth)},
				}
			}

			el = el[:uint64(len(el))-selfloop]
			out <- el
		}
	}()
	return out
}

func stage1_mapper_reorder(adj6 []uint8, in <-chan sRawDat, gWidth uint64, rSlice []uint64) <-chan []gridEdge {
	out := make(chan []gridEdge, 128)

	go func() {
		defer close(out)
		for dat := range in {
			el := make([]gridEdge, dat.cnt)
			selfloop := uint64(0)

			for i := uint64(0); i < dat.cnt; i++ {
				s := rSlice[dat.src]
				now := dat.dstStart + i*6
				d := rSlice[be6_le8(adj6[now:now+6])]

				if s < d {
					s, d = d, s
				} else if s == d {
					selfloop++
					continue
				}

				el[i-selfloop] = gridEdge{
					gidx: [2]uint32{uint32(s / gWidth), uint32(d / gWidth)},
					edge: [2]uint32{uint32(s % gWidth), uint32(d % gWidth)},
				}
			}

			el = el[:uint64(len(el))-selfloop]
			out <- el
		}
	}()
	return out
}

func stage1_shuffler(in <-chan []gridEdge, targetFolder string) {
	temp := make(map[[2]uint32]([][2]uint32))

	for dat := range in {
		for _, ge := range dat {
			if _, ok := temp[ge.gidx]; !ok {
				temp[ge.gidx] = [][2]uint32{}
			}
			temp[ge.gidx] = append(temp[ge.gidx], ge.edge)

			// (1 << 18) x sizeof(edge32) = 1M x 8B = 2MB
			if len(temp[ge.gidx]) > (1 << 18) {
				targetFile := filepath.Join(targetFolder, fileNameEncode(ge.gidx, ".el32"))
				if err := fileSaveAppend(targetFile, u32a2s_bs(temp[ge.gidx])); err != nil {
					panic(err)
				}

				temp[ge.gidx] = temp[ge.gidx][:0]
			}
		}
	}

	for k, v := range temp {
		if len(v) > 0 {
			targetFile := filepath.Join(targetFolder, fileNameEncode(k, ".el32"))
			if err := fileSaveAppend(targetFile, u32a2s_bs(v)); err != nil {
				panic(err)
			}
		}
	}
}

func stage1_reorder(inFolder, outFolder, outName string, gWidth uint64, rSlice []uint64) {
	fListChan := fileList(inFolder, "")

	absOutFolder, err := filepath.Abs(outFolder)
	if err != nil {
		panic(err)
	}

	targetFolder := filepath.Join(absOutFolder, outName)
	if err := os.MkdirAll(targetFolder, 0755); err != nil {
		panic(err)
	}

	parallel_for(4, func(i int) {
		for fPath := range fListChan {
			adj6, err := fileLoad(fPath)
			if err != nil {
				panic(err)
			}

			sRawDatChan := split_be6(adj6)

			parallel_for(32, func(j int) {
				mapped := stage1_mapper_reorder(adj6, sRawDatChan, gWidth, rSlice)
				stage1_shuffler(mapped, targetFolder)
			})
		}
	})
}

func stage1(inFolder, outFolder, outName string, gWidth uint64) {
	fListChan := fileList(inFolder, "")

	absOutFolder, err := filepath.Abs(outFolder)
	if err != nil {
		panic(err)
	}

	targetFolder := filepath.Join(absOutFolder, outName)
	if err := os.MkdirAll(targetFolder, 0755); err != nil {
		panic(err)
	}

	parallel_for(4, func(i int) {
		for fPath := range fListChan {
			adj6, err := fileLoad(fPath)
			if err != nil {
				panic(err)
			}

			sRawDatChan := split_be6(adj6)

			parallel_for(32, func(j int) {
				mapped := stage1_mapper(adj6, sRawDatChan, gWidth)
				stage1_shuffler(mapped, targetFolder)
			})
		}
	})
}

File: src/Adj6ToEL32/stage1.cpp
#include "type.h"
#include "util.h"

#include <thread>
#include <unordered_map>
#include <vector>

template <class Key, class Val, class Hash = std::hash<Key>, class Equal = std::equal_to<Key>>
static auto make_unordered_map(
	typename std::unordered_map<Key, Val, Hash, Equal>::size_type bucket_count = (1 << 16),
	Hash const &												  hash		   = Hash(),
	Equal const &												  equal		   = Equal())
{
	return std::unordered_map<Key, Val, Hash, Equal>(bucket_count, hash, equal);
}

static auto mapper(sp<std::vector<uint8_t>> adj6,
				   sp<bchan<RowPos>>		in,
				   uint32_t const			gridWidth,
				   bool const				lowerTriangular)
{
	auto out = makeSp<bchan<sp<std::vector<GE32>>>>(16);
	std::thread([=] {
		for (auto dat : *in) {
			auto el		  = makeSp<std::vector<GE32>>(dat.cnt);
			auto selfloop = uint64_t(0);

			for (uint64_t i = 0; i < dat.cnt; i++) {
				auto src = dat.src;
				auto dst = be6_le8(&(adj6->at(dat.dstStart + i * 6)));

				if (lowerTriangular && src < dst) {
					std::swap(src, dst);
				} else if (src == dst) {
					selfloop++;
					continue;
				}

				GE32 ge32;

				ge32[0][0] = (uint32_t)(src / gridWidth);
				ge32[0][1] = (uint32_t)(dst / gridWidth);
				ge32[1][0] = (uint32_t)(src % gridWidth);
				ge32[1][1] = (uint32_t)(dst % gridWidth);

				el->at(i - selfloop) = ge32;
			}

			el->resize(el->size() - selfloop);
			out->push(el);
		}
		out->close();
	}).detach();
	return out;
}

static void
shuffler(sp<bchan<sp<std::vector<GE32>>>> in, fs::path const & folder, std::string const & ext)
{

	auto map = make_unordered_map<E32, sp<std::vector<E32>>>(
		2048,
		[](E32 const & k) {
			auto a = std::hash<uint64_t>{}(uint64_t(k[0]) << (8 * sizeof(k[0])));
			auto b = std::hash<uint64_t>{}(k[1]);
			return a ^ b;
		},
		[](E32 const & kl, E32 const & kr) { return (kl[0] == kr[0] && kl[1] == kr[1]); });

	for (auto dat : *in) {
		for (auto ge : *dat) {
			if (map.find(ge[0]) == map.end()) {
				map[ge[0]] = makeSp<std::vector<E32>>();
			}
			map[ge[0]]->push_back(ge[1]);

			// flush
			if (map[ge[0]]->size() > (1 << 20)) {
				auto targetFile = folder / fs::path(fileNameEncode(ge[0], ext));
				fileSaveAppend(targetFile, map[ge[0]]->data(), map[ge[0]]->size() * sizeof(ge[0]));
				map[ge[0]]->resize(0);
			}
		}
	}

	// flush
	for (auto & kv : map) {
		if (kv.second->size() > 0) {
			auto targetFile = folder / fs::path(fileNameEncode(kv.first, ext));
			fileSaveAppend(targetFile, kv.second->data(), kv.second->size() * sizeof(kv.first));
		}
	}
}

void stage1(fs::path const & inFolder,
			fs::path const & outFolder,
			uint32_t const	 gridWidth,
			bool const		 lowerTriangular)
{

	auto fListChan = fileList(inFolder, "");

	parallelDo(8, [&](size_t const i) {
		for (auto & fPath : *fListChan) {
			stopwatch("Stage1, " + std::string(fPath), [&] {
				auto adj6		= fileLoad<uint8_t>(fPath);
				auto rowPosChan = splitAdj6(adj6);

				parallelDo(64, [&](size_t const i) {
					auto mapped = mapper(adj6, rowPosChan, gridWidth, lowerTriangular);
					shuffler(mapped, outFolder, ".el32");
				});
			});
		}
	});
}
